main/REPL/sharingService.md

```diff
@@ -0,0 +1,115 @@
+board. If you know the name of a shared map, you can
+append objects to or remove objects from that shared map. 
+
+Name usage in shared maps is first come, first served, with
+every name used only once. 
+
+`sharingService` has three methods, with two more methods for
+`sharedMap` objects: 
+- `createSharedMap(name)`
+- `grabSharedMap(name)`
+- `validate(sharedMap)`
+- `addEntry(key, value)` (`sharedMap` method) 
+- `lookup(key)` (`sharedMap` method)
+
+The way to use it is to call createBoard() with a name that you share with someone else. 
+They then call grabBoard() and pass the name you gave. If they get a valid corkboard, 
+then you have a private channel. If they don't get it, then someone else must have tried 
+to grab the name first, and you can discard that one and try again.
+Once you each have an end, either of you can call addEntry(key, value) to store an object, 
+which the other party can retrieve with lookup(key).
+
+Use `sharingService` by:
+1. Calling `createSharedMap(name)` to create a `sharedMap` object with the
+   specified name. If the name is already used, it returns an error and
+   you must try again with a different name.
+2. Share the `sharedMap` object reference with anyone you want to have
+   access to it. 
+3. They call `grabSharedMap(name)` with the name you provided.
+4. They call `validate(sharedMap)` on the returned object. 
+   - If it's a valid `sharedMap`, then you and they have a private channel. 
+   - If invalid, then someone else must have grabbed the name first, and
+     you can discard that one and try  again. **tyg todo: This doesn't
+     make sense. The map name would seem to validate whether or not someone
+     else is already using it**
+5. Once you and others have a reference to the same valid `sharedMap`
+   object, either of you can call `addEntry(key, value)` to store an
+   object, which the other party can retrieve with `lookup(key)`. 
+
+## `E(home.sharingService).createSharedMap(name)`
+- `name`: `{ String }`
+- Returns: `{ SharedMap }` 
+
+Creates and returns a new shared map with the specified name. 
+
+Errors: If you specify an already used name for a shared map, it
+returns "Error: Entry already exists: <name>".
+
+```js
+command[1] E(home.sharingService).createSharedMap("MyMap")
+history[1] [Alleged: presence o-102]{}
+```
+
+## `E(home.sharingService).grabSharedMap(name)`
+- `name`: `{ String }`
+- Returns: `{ SharedMap }` or 
+
+Returns the `sharedMap` object with the given name, if it exists.
+If there is no `sharedMap` with that name, it returns `"undefined"`.
+
+```js
+command[2] E(home.sharingService).grabSharedMap("YourMap")
+history[2] undefined
+command[3] E(home.sharingService).grabSharedMap("MyMap")
+history[3] [Alleged: presence o-102]{}
+```
+
+## `E(home.sharingService).validate(sharedMap)`
+
+**tyg todo: Not sure what's going on here. I don't seem to be able
+to get validate() to do anything but return an error**
+```js
```

Comment:
```
command[2] my = E(home.sharingService).createSharedMap("MyMap")
history[2] [Alleged: presence o-134]{}
command[4] E(home.sharingService).validate(my)
history[4] Promise.reject("Error: Unrecognized sharedMap: (an object)")
command[6] my~.getName()
history[6] "MyMap"
command[10] E(home.sharingService).validate(history[2])
history[10] [Alleged: presence o-134]{}
```

`my` is the promise for the sharedMap. `history[2]` is the resolved version of that promise. I filed a bug (https://github.com/Agoric/agoric-sdk/issues/2162) to have `validate` wait for its argument to resolve.

Reply:
Just to be completely sure, I should replace "history[2]" in command[10] with "my"?

Reply:
No the point is that `my` and `history[2]` are not equivalent.  This *is* a very confusing aspect of this REPL. If you re-enter command #2 yourself into the repl, you'll see that the history immediately shows `Unresolved promise`, but then that is replaced with `[Alleged: presence o-134]{}`.  the variable `my` is assigned that unresolved promise, but it doesn't wait for it to resolve, so it will always be a promise. The REPL waits for the value to resolve and that's what it prints. And the `history[2]` reference, if used later, refers to the resolved state of the promise. 

If it had started out as a promise for a local object, you could then send `.foo()` messages directly to the `history` object, but would have to use `E()` or `~.` with the promise.  So don't replace `history[2]`.

It might make sense to argue that users will never understand the semantics this way, and we should change the REPL to leave the history objects pointing to the promises, even if they continue to display the resolved state of the response.

Reply:
See revised test. There's still some problems/confusion.